### DISCUSSION TITLE
"System.TypeInitializationException : The type initializer for '<Module>' threw an exception" when Realm Database NuGet Package is installed

### DISCUSSION THREAD

**Discussion by HobDev** [xunit/xunit#2361] [Question] [answered]
I am working on Xamarin.Forms project. I have a xUnit test project targeting dotnetcore 3.1. When I run the tests I get the error mentioned in the title. : The complete test detail summary is as follows:

> Milk.UnitTests.LoginTest.SelectedLanguageApplied(language: "Hindi")
>    Source: LoginTests.cs line 22
>    Duration: < 1 ms
> 
>   Message: 
>     System.TypeInitializationException : The type initializer for '<Module>' threw an exception.
>     ---- System.InvalidProgramException : The JIT compiler encountered invalid IL code or an internal limitation.
> 
>   Stack Trace: 
>     LoginTest.SelectedLanguageApplied(String language)
>     AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine)
>     LoginTest.SelectedLanguageApplied(String language)
>     ----- Inner Stack Trace -----
>     cctor()

The test method is:
```
[Theory]
        [InlineData("Hindi")]
        [InlineData("English")]

        //The choosen language is applied
        public async void SelectedLanguageApplied(string language)
        {
            //Arrange
            LoginViewModel loginViewModel = new LoginViewModel();

            //Act
            string result = await loginViewModel.ChangeLanguage(language);

            //Assert
            Assert.Equal(language, result);
        }
```

The ViewModel method of the above test is:
    

```
 public async Task<string> ChangeLanguage(string selectedLanguage)
        {
            _selectedLanguage = selectedLanguage;
            CultureInfo language = CultureInfo.GetCultures(CultureTypes.NeutralCultures).ToList().First(element => element.EnglishName.Contains(selectedLanguage));
            Thread.CurrentThread.CurrentUICulture = language;
            AppResources.Culture = language;
            return language.EnglishName;
           
        }
```

I am really frustated as the tests was running successfully before. 
Thank you

**Comment by bradwilson** [comment]
Given the code above, I'm unable to repro your problem:

`cat .\xunit-2361.csproj`

```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <TargetFramework>netcoreapp3.1</TargetFramework>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="Microsoft.NET.Test.Sdk" Version="16.11.0" />
    <PackageReference Include="xunit" Version="2.4.1" />
    <PackageReference Include="xunit.runner.visualstudio" Version="2.4.3">
      <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
      <PrivateAssets>all</PrivateAssets>
    </PackageReference>
  </ItemGroup>

</Project>
```

`cat .\UnitTest1.cs`

```csharp
using System.Globalization;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;
using Xunit;

public class LoginViewModel
{
    public async Task<string> ChangeLanguage(string selectedLanguage)
    {
        CultureInfo language = CultureInfo.GetCultures(CultureTypes.NeutralCultures).ToList().First(element => element.EnglishName.Contains(selectedLanguage));
        Thread.CurrentThread.CurrentUICulture = language;
        return language.EnglishName;
    }
}

public class UnitTest1
{
    [Theory]
    [InlineData("Hindi")]
    [InlineData("English")]
    public async void SelectedLanguageApplied(string language)
    {
        LoginViewModel loginViewModel = new LoginViewModel();
        string result = await loginViewModel.ChangeLanguage(language);
        Assert.Equal(language, result);
    }
}
```

`dotnet test`

```
  Determining projects to restore...
  All projects are up-to-date for restore.
  xunit-2361 -> C:\Dev\repro\xunit-2361\bin\Debug\netcoreapp3.1\xunit-2361.dll
Test run for C:\Dev\repro\xunit-2361\bin\Debug\netcoreapp3.1\xunit-2361.dll (.NETCoreApp,Version=v3.1)
Microsoft (R) Test Execution Command Line Tool Version 16.10.0
Copyright (c) Microsoft Corporation.  All rights reserved.

Starting test execution, please wait...
A total of 1 test files matched the specified pattern.

Passed!  - Failed:     0, Passed:     2, Skipped:     0, Total:     2, Duration: 2 ms - xunit-2361.dll (netcoreapp3.1)
```

A full repro project will be necessary.

**Reply by HobDev** [reply]
@bradwilson This only happens when there is [Realm Database](https://docs.mongodb.com/realm/get-started/introduction-mobile/) used in the app. If I Remove Realm Models along with all the realm related code the test runs and pass successfully. Here is the [link](https://drive.google.com/file/d/1wKpq6GFvbOXmZq_KogrJAx6mhUX33AqO/view) to sample project. 
Realm and Unit tests are both required in the project. So, some how they need to work together.

**Comment by HobDev** [comment, marked as answer]
The error has gone by removing the Realm Package from the Shared Project(code tested from this project). The realm package already installed in the Models project. which have all the Models of the App. The Shared project reference the Model project so it can still work with Realm.
